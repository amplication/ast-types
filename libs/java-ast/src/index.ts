// Core exports
export * from "./lib/core/AstNode";
export * from "./lib/core/Writer";

// AST nodes exports
export * from "./lib/ast/Access";
export * from "./lib/ast/Annotation";
export * from "./lib/ast/Class";
export * from "./lib/ast/ClassReference";
export * from "./lib/ast/CodeBlock";
export * from "./lib/ast/Enum";
export * from "./lib/ast/Field";
export * from "./lib/ast/GenericType";
export * from "./lib/ast/Import";
export * from "./lib/ast/Interface";
export * from "./lib/ast/Method";
export * from "./lib/ast/Parameter";
export * from "./lib/ast/Type";

// Utils exports
// (Add utility functions as needed)
