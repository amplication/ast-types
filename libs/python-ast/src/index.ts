export * from "./lib/ast";
export * from "./lib/core";
