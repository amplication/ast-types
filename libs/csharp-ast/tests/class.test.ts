import { CsharpSupport, CsharpSupport as csharp } from "../src/";

describe("class", () => {
  it("basic", () => {
    const myClass = csharp.class_({
      name: "MyClassName",
      namespace: "Amplication",
      partial: true,
      access: "public",
    });
    myClass.addField(
      csharp.field({
        name: "watered",
        type: csharp.Types.boolean(),
        access: "public",
        get: true,
        set: true,
        annotations: [
          csharp.annotation({
            reference: csharp.classReference({
              name: "JsonProperty",
              namespace: "System.Text.Json.Serialization",
            }),
            argument: '"watered"',
          }),
          csharp.annotation({
            reference: csharp.classReference({
              name: "JsonAttribute",
              namespace: "System.Text.Json.Serialization",
            }),
            argument: '"JsonAttribute"',
          }),
        ],
      })
    );
    expect(myClass.name).toBe("MyClassName");
    expect(myClass.toString()).toMatchSnapshot();
  });

  it("complex", () => {
    const myClass = csharp.class_({
      name: "MyClassName",
      namespace: "Amplication",
      partial: true,
      access: "public",
      abstract: true,
    });
    myClass.addField(
      csharp.field({
        name: "watered",
        type: csharp.Types.boolean(),
        access: "public",
        get: true,
        init: true,
        annotations: [
          csharp.annotation({
            reference: csharp.classReference({
              name: "JsonProperty",
              namespace: "System.Text.Json.Serialization",
            }),
            argument: '"watered"',
          }),
        ],
        summary: "Whether the myClass has been watered",
      })
    );

    myClass.addField(
      CsharpSupport.field({
        name: "_myField",
        access: "protected",
        readonly_: true,
        type: CsharpSupport.Types.string(),
      })
    );

    myClass.addMethod(
      csharp.method({
        access: "public",
        name: "MyVoidMethod",
        isAsync: true,
        return_: csharp.Types.genericReference(
          CsharpSupport.genericClassReference({
            reference: csharp.classReference({
              name: "ActionResult",
              namespace: "Microsoft.AspNetCore.Mvc",
            }),
            innerType: CsharpSupport.Types.list(
              CsharpSupport.Types.reference(
                CsharpSupport.classReference({
                  name: `myDto`,
                  namespace: `MyService.APIs.Dtos`,
                })
              )
            ),
          })
        ),
        parameters: [
          csharp.parameter({
            name: "simpleParameter",
            type: csharp.Types.integer(),
          }),
          csharp.parameter({
            name: "listParameter",
            type: csharp.Types.list(csharp.Types.string()),
          }),
          csharp.parameter({
            name: "optionalParameter",
            type: csharp.Types.optional(csharp.Types.string()),
          }),
          csharp.parameter({
            name: "parameterWithAnnotation",
            type: csharp.Types.integer(),
            annotations: [
              csharp.annotation({
                reference: csharp.classReference({
                  name: "FromQuery",
                  namespace: "Microsoft.AspNetCore.Mvc",
                }),
              }),
            ],
          }),
        ],
        body: csharp.codeblock({
          code: 'Console.WriteLine(@"Received these: {simpleParameter} {listParameter} {optionalParameter}");',
        }),
      })
    );

    myClass.addMethod(
      csharp.method({
        access: "public",
        name: "MyMethod",
        isAsync: true,
        return_: csharp.Types.integer(),
        parameters: [
          csharp.parameter({
            name: "simpleParameter",
            type: csharp.Types.integer(),
          }),
          csharp.parameter({
            name: "listParameter",
            type: csharp.Types.list(csharp.Types.string()),
          }),
          csharp.parameter({
            name: "optionalParameter",
            type: csharp.Types.optional(csharp.Types.string()),
          }),
        ],
        body: csharp.codeblock({
          code: 'Console.WriteLine(@"Received these: {simpleParameter} {listParameter} {optionalParameter}");',
        }),
        annotations: [
          csharp.annotation({
            reference: csharp.classReference({
              name: "HttpGet",
              namespace: "Microsoft.AspNetCore.Mvc",
            }),
            argument: "{Id}",
          }),
          csharp.annotation({
            reference: csharp.classReference({
              name: "Authorize",
              namespace: "Microsoft.AspNetCore.Authorization;",
            }),
            argument: 'Roles = "admin,user"',
          }),
        ],
      })
    );
    expect(myClass.name).toBe("MyClassName");
    expect(myClass.toString()).toMatchSnapshot();
  });

  it("enum", () => {
    const myEnum = csharp.enum_({
      name: "MyEnum",
      namespace: "Amplication",
      access: "public",
      annotations: [
        csharp.annotation({
          reference: csharp.classReference({
            name: "JsonConverter",
            namespace: "System.Text.Json.Serialization",
          }),
          argument: "typeof(StringEnumConverter)",
        }),
      ],
    });
    myEnum.addMember({
      name: "Value1",
      value: "value1",
    });
    myEnum.addMember({
      name: "Value2",
      value: "value2",
    });

    expect(myEnum.name).toBe("MyEnum");
    expect(myEnum.toString()).toMatchSnapshot();
  });

  it("is abstract", () => {
    const myClass = csharp.class_({
      name: "MyClassName",
      namespace: "Amplication",
      partial: true,
      abstract: true,
      access: "public",
    });
    expect(myClass.name).toBe("MyClassName");
    expect(myClass.toString()).toMatchSnapshot();
  });
  it("is static", () => {
    const myClass = csharp.class_({
      name: "MyClassName",
      namespace: "Amplication",
      static_: true,
      access: "public",
    });
    expect(myClass.name).toBe("MyClassName");
    expect(myClass.toString()).toMatchSnapshot();
  });
  it.each`
    modifier    | otherModifier
    ${"sealed"} | ${"abstract"}
    ${"static"} | ${"abstract"}
    ${"sealed"} | ${"static"}
  `(
    "is throw if $modifier and $otherModifier are both true",
    ({ modifier, otherModifier }) => {
      const myClass = csharp.class_({
        name: "MyClassName",
        namespace: "Amplication",
        static_: modifier === "static" || otherModifier === "static",
        sealed: modifier === "sealed" || otherModifier === "sealed",
        abstract: modifier === "abstract" || otherModifier === "abstract",
        access: "public",
      });

      expect(() => myClass.toString()).toThrow(
        `A class can only be one of abstract, sealed, or static at a time`
      );
    }
  );
});
