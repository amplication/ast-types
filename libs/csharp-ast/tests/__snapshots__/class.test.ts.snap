// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class basic 1`] = `
"using System.Text.Json.Serialization;

namespace Amplication;

public partial class MyClassName
{
    [JsonProperty("watered")]
    [JsonAttribute("JsonAttribute")]
    public bool watered { get; set; }
}
"
`;

exports[`class complex 1`] = `
"using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Mvc;
using MyService.APIs.Dtos;
using Microsoft.AspNetCore.Authorization;;

namespace Amplication;

public abstract partial class MyClassName
{
    protected readonly string _myField;

    /// <summary>
    /// Whether the myClass has been watered
    /// </summary>
    [JsonProperty("watered")]
    public bool watered { get; init; }

    public async Task<ActionResult<List<myDto>>> MyVoidMethod(int simpleParameter, List<string> listParameter, string? optionalParameter, [FromQuery()]
    int parameterWithAnnotation) {
        Console.WriteLine(@"Received these: {simpleParameter} {listParameter} {optionalParameter}");}

    [HttpGet({Id})]
    [Authorize(Roles = "admin,user")]
    public async Task<int> MyMethod(int simpleParameter, List<string> listParameter, string? optionalParameter) {
        Console.WriteLine(@"Received these: {simpleParameter} {listParameter} {optionalParameter}");}

}
"
`;

exports[`class enum 1`] = `
"using System.Text.Json.Serialization;
using System.Runtime.Serialization;

namespace Amplication;

[JsonConverter(typeof(StringEnumConverter))]
public enum MyEnum
{
    [EnumMember(Value = "value1")]
    Value1,

    [EnumMember(Value = "value2")]
    Value2
}
"
`;

exports[`class is abstract 1`] = `
"namespace Amplication;

public abstract partial class MyClassName
{
}
"
`;

exports[`class is static 1`] = `
"namespace Amplication;

public static class MyClassName
{
}
"
`;
